# Lab 01 

In [1]:
# 1.1
# Import the package and read the data
import pandas as pd
covid = pd.read_csv("covid-scandinavia.csv", sep = '\t')
covid.head()

code2  country  state        date       type  count    lockdown  population  \
0    DK  Denmark    NaN  2020-01-22  Confirmed      0  2020-03-11     5837213   
1    DK  Denmark    NaN  2020-01-22     Deaths      0  2020-03-11     5837213   
2    DK  Denmark    NaN  2020-01-23  Confirmed      0  2020-03-11     5837213   
3    DK  Denmark    NaN  2020-01-23     Deaths      0  2020-03-11     5837213   
4    DK  Denmark    NaN  2020-01-24  Confirmed      0  2020-03-11     5837213   

   countPC  growth  growthPC  
0      0.0     NaN       NaN  
1      0.0     NaN       NaN  
2      0.0     0.0       0.0  
3      0.0     0.0       0.0  
4      0.0     0.0       0.0

In [2]:
# 1.2
# Report the number of rows and columns in the dataset
covid.shape

(6528, 11)

There are 6528 rows and 11 columns for this dataset.

In [3]:
# 1.3 
# Report the variable names along with the data type of each variable
covid.dtypes

code2          object
country        object
state         float64
date           object
type           object
count           int64
lockdown       object
population      int64
countPC       float64
growth        float64
growthPC      float64
dtype: object

Based on the information, the dataset has the following variables:
code2: object (assumed to be a two-letter country code);
country: object (name of the country);
state: float64 (assumed to be a code for a state or province, but can be missing);
date: object (date of the observation in yyyy-mm-dd format);
type: object (type of the observation, e.g., cases or deaths);
count: int64 (number of cases or deaths);
lockdown: object (assumed to be a binary variable indicating if a lockdown was in effect on the observation date);
population: int64 (estimated population of the region);
countPC: float64 (assumed to be the count per capita);
growth: float64 (assumed to be the growth rate of the count compared to the previous day);
growthPC: float64 (assumed to be the growth rate per capita);
The dtypes method provides information on the data type of each column in the DataFrame.

In [4]:
# 1.4 
# Create a new sub-dataframe that contains all observations but only variables country, date, type, count, and population.
sub_dataframe = covid[["country", "date", "type", "count", "population"]]
print("Number of rows and columns:", sub_dataframe.shape)
print(sub_dataframe.head())

Number of rows and columns: (6528, 5)
   country        date       type  count  population
0  Denmark  2020-01-22  Confirmed      0     5837213
1  Denmark  2020-01-22     Deaths      0     5837213
2  Denmark  2020-01-23  Confirmed      0     5837213
3  Denmark  2020-01-23     Deaths      0     5837213
4  Denmark  2020-01-24  Confirmed      0     5837213


In [5]:
# 2.1
# create a subset with only confirmed cases
confirmed = sub_dataframe[sub_dataframe['type'] == 'Confirmed']
confirmed

country        date       type    count  population
0     Denmark  2020-01-22  Confirmed        0     5837213
2     Denmark  2020-01-23  Confirmed        0     5837213
4     Denmark  2020-01-24  Confirmed        0     5837213
6     Denmark  2020-01-25  Confirmed        0     5837213
8     Denmark  2020-01-26  Confirmed        0     5837213
...       ...         ...        ...      ...         ...
6518   Sweden  2022-04-12  Confirmed  2491980    10377781
6520   Sweden  2022-04-13  Confirmed  2491980    10377781
6522   Sweden  2022-04-14  Confirmed  2495996    10377781
6524   Sweden  2022-04-15  Confirmed  2495996    10377781
6526   Sweden  2022-04-16  Confirmed  2495996    10377781

[3264 rows x 5 columns]

There are 3264 cases in the subset.

In [6]:
confirmed.shape

(3264, 5)

In [10]:
# 2.2
# create a new column for confirmed cases per capita
confirmed['cases_per_capita'] = confirmed['count'] / confirmed['population']
confirmed

/tmp/ipykernel_137/2928224583.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  confirmed['cases_per_capita'] = confirmed['count'] / confirmed['population']


country        date       type    count  population  cases_per_capita
0     Denmark  2020-01-22  Confirmed        0     5837213          0.000000
2     Denmark  2020-01-23  Confirmed        0     5837213          0.000000
4     Denmark  2020-01-24  Confirmed        0     5837213          0.000000
6     Denmark  2020-01-25  Confirmed        0     5837213          0.000000
8     Denmark  2020-01-26  Confirmed        0     5837213          0.000000
...       ...         ...        ...      ...         ...               ...
6518   Sweden  2022-04-12  Confirmed  2491980    10377781          0.240126
6520   Sweden  2022-04-13  Confirmed  2491980    10377781          0.240126
6522   Sweden  2022-04-14  Confirmed  2495996    10377781          0.240513
6524   Sweden  2022-04-15  Confirmed  2495996    10377781          0.240513
6526   Sweden  2022-04-16  Confirmed  2495996    10377781          0.240513

[3264 rows x 6 columns]

In [8]:
# 2.3
confirmed.count / confirmed.population

TypeError: unsupported operand type(s) for /: 'method' and 'int'


The expression data.count/data.population results in an type error because count is a method in pandas DataFrames and Series, and it is not intended to be used as a column name. To get around this problem, we can use the square bracket notation to access the 'count' column, like this: data['count']/data['population'].

In [15]:
# 2.4
# create a new DataFrame for confirmed cases and copy the subset into it
confirmed_cases = sub_dataframe[sub_dataframe['type'] == 'Confirmed'].copy()
confirmed_cases['cases_per_capita'] = confirmed_cases['count'] / confirmed_cases['population']
confirmed_cases['cases_per_capita']

0       0.000000
2       0.000000
4       0.000000
6       0.000000
8       0.000000
          ...   
6518    0.240126
6520    0.240126
6522    0.240513
6524    0.240513
6526    0.240513
Name: cases_per_capita, Length: 3264, dtype: float64

In [12]:
# 2.5
# extract the cases per capita variable for Sweden and Norway
sweden_cases = confirmed_cases[confirmed_cases['country'] == 'Sweden']['cases_per_capita']
norway_cases = confirmed_cases[confirmed_cases['country'] == 'Norway']['cases_per_capita']

In [13]:
sweden_cases

4896    0.000000
4898    0.000000
4900    0.000000
4902    0.000000
4904    0.000000
          ...   
6518    0.240126
6520    0.240126
6522    0.240513
6524    0.240513
6526    0.240513
Name: cases_per_capita, Length: 816, dtype: float64

In [14]:
norway_cases

3264    0.000000
3266    0.000000
3268    0.000000
3270    0.000000
3272    0.000000
          ...   
4886    0.263276
4888    0.263425
4890    0.263496
4892    0.263548
4894    0.263625
Name: cases_per_capita, Length: 816, dtype: float64

In [ ]:
# 2.6
# check the data structure of the newly created variables
print("Data structure of sweden_cases:", type(sweden_cases))
print("Data structure of norway_cases:", type(norway_cases))

### 2.7
The index of these Series is currently the row number of each observation in the DataFrame. While this is a valid index, it is not particularly useful for analyzing time series data. A better index would be the date of each observation, as it would allow us to easily compare the number of cases per capita over time.Currently, the index of these Series is the row number of each DataFrame observation. This is a legitimate index, although it is not especially useful for time series data analysis. The date of each observation would be a superior indicator since it would allow us to readily compare the number of instances per capita across time.

In [ ]:
# 2.8
# set the index of each series to the date column
sweden_cases.index = confirmed_cases[confirmed_cases['country'] == 'Sweden']['date']
norway_cases.index = confirmed_cases[confirmed_cases['country'] == 'Norway']['date']

# print the first few rows to check the new index
print("Sweden cases per capita:")
print(sweden_cases.head())
print("\nNorway cases per capita:")
print(norway_cases.head())

In [ ]:
# 3.1
# Method 1: print the confirmed cases per capita in Sweden and Norway for different dates by using .loc
print(sweden_cases.loc['2020-05-01'])
print(norway_cases.loc['2020-05-01'])

print(sweden_cases.loc['2020-12-01'])
print(norway_cases.loc['2020-12-01'])

print(sweden_cases.loc['2021-07-01'])
print(norway_cases.loc['2021-07-01'])

print(sweden_cases.loc['2022-01-01'])
print(norway_cases.loc['2022-01-01'])

In [ ]:
# Method 2: print the confirmed cases per capita in Sweden and Norway for different dates by using index
print(norway_cases[norway_cases.index == '2020-05-01'])
print(norway_cases[norway_cases.index == '2020-12-01'])
print(norway_cases[norway_cases.index == '2021-07-01'])
print(norway_cases[norway_cases.index == '2022-01-01'])

print(sweden_cases[sweden_cases.index == '2020-05-01'])
print(sweden_cases[sweden_cases.index == '2021-07-01'])
print(sweden_cases[sweden_cases.index == '2020-05-01'])
print(sweden_cases[sweden_cases.index == '2020-05-01'])

According to the output, confirmed cases per capita in Sweden were typically greater than in Norway at each of the four time periods. This might imply that Sweden's "hands-off" attitude may have had some effect on the virus's spread in the country. However, it is crucial to emphasize that many other factors may have contributed to the discrepancies between Sweden and Norway, and a more thorough research would need to investigate a broader variety of variables and data sources.